### DISCUSSION TITLE
SimulationController

### DISCUSSION THREAD

**Discussion by quasiquote1** [sparcians/map#271] [Q&A] [answered]
Hi,

I looked through ExampleSimulation.cpp in the CoreModel, and noticed that there's a class called ExampleController which inherits from the SimulationController (https://sparcians.github.io/map/classsparta_1_1app_1_1Simulation_1_1SimulationController.html#details).

I do not see any examples of where this is being used in the Yaml files.  Can you please explain its utility and some examples of this callback?

Separately, is this the only way to interact with the Sparta simulation, or are there other hooks (for example through Python APIs) that are implemented as well?

Thanks for your help!

**Comment by klingaard** [comment, marked as answer]
Actually there is an example of how it's used here: https://github.com/sparcians/map/blob/master/sparta/example/CoreModel/ctrl.yaml

The history of this controller came from the development of a python shell, admittedly still in its infancy.   The concept allows a user of the model to register a specific given control `(pause,` `resume,` `term)` with an event in the simulation, say a `sparta::Counter` hitting a certain value.  You can also define custom controls with your own meanings (like `eat` and `sleep` in the example above).

However, it extends beyond the python shell.  You can use this control on the command line to, for example, stop simulation when a specific trigger happens.  Kinda handy when debugging your simulation.  As an example, let's say you want to dump a log of a run, but simulation should stop on the 10th ALU instruction issued.  You can create a `control.yaml` file to stop simulation:
```
# control.yaml
content:
  term:   'cpu.core0.alu0.stats.total_insts_issued >= 10'
```
Change the code in [ExampleSimulation.cpp](https://github.com/sparcians/map/blob/b9fe36612350b49d94fbd8f46a79df0222d23d1a/sparta/example/CoreModel/src/ExampleSimulation.cpp#L1197) to call `sim->getScheduler()->stopRunning();`  (need to cache a copy of the `sim` object) and then run the simulator:
```
sparta_core_example -l top debug 1 --control control.yaml
```

**Reply by klingaard** [reply]
On another note, I'll update the `SimulationController` documentation.  Thanks for pointing that out.

**Comment by quasiquote1** [comment]
Hi,
I recently tried to add my own custom SimulationController, and noticed it was very finicky in whether my control.yaml was being invoked (and thus whether my callbacks occurred).  My custom trigger expressions were similar to the one you used in your example, except I was using a different sparta::Counter (and comparison values).

I noticed that sometimes if I incremented the sparta::Counter in a certain routine in the Tick phase, my callbacks would occur, but if I incremented the sparta::Counter in a different routine scheduled in the next Cycle it would *not* occur.

Are there any known restrictions that might prevent the trigger expressions from being evaluated correctly, due to how sparta::Counter objects are incremented?

Thanks in advance!

**Reply by klingaard** [reply]
Hmmm... not sure.  The different routine -- anything special about it?  Is it called from an event on different clock?  On a different phase?

**Reply by quasiquote1** [reply]
No, they were both functions in the Tick phase and using the same Clock.

**Reply by klingaard** [reply]
Ok, I understand the use case now.

If I remember how Colby implemented this back in the day, a `sparta::trigger::ContextCounterTrigger` (which is what is used under the covers) will be enabled or placed on the sparta scheduler to be called until the condition they are requested to execute has been met.  

When that happens, the trigger fires, a callback is called and the counter is no longer watched.  This is documented (but not in the Doxygen -- I need to enable it) [here](https://github.com/sparcians/map/blob/13582f0faa621282ff6b1841fa1f2df95ba407ef/sparta/sparta/trigger/ContextCounterTrigger.hpp#L24-L41).

So, I'd expect you to receive a call once the condition is met, but then it will go silent.  _Technically_ you can create your own `sparta::trigger::ContextCounterTrigger` for what you might be trying to do.  After relying on Sparta to fire the first callback, you can create a new one for another condition, say when a counter reaches the next 100K value.

Hope this helps.
